### DISCUSSION TITLE
Adding Blocks on the fly

### DISCUSSION THREAD

**Discussion by traustitj** [PiranhaCMS/piranha.core#2082] [Q&A]
This is a code I have been trying to get working

        public async Task<IActionResult> FrontPage(Guid id, bool draft = false)
        {
            try
            {
                var model = await _loader.GetPageAsync<FrontPage>(id, HttpContext.User, draft);
                
                // Todo: Find out how to get this working
                
                //var pages = await _api.Pages.GetAllAsync<Page>();

                //var interests = pages.Where(x => x.Details.Interesting.Value == true).Take(3).ToList();
                //var columns = new ColumnBlock();
                //foreach (var interest in interests)
                //{
                //    var pageBlock = new Piranha.Extend.Blocks.PageBlock();
                //    pageBlock.Body = interest;
                //    columns.Items.Add(pageBlock);
                //}

                return View(model);
            }
            catch (UnauthorizedAccessException)
            {
                return Unauthorized();
            }
        }

The commented out code should simply add blocks to the page, like if you want to find the most popular pages, to show them as page links. The code works but does not do what I would expect.

The thing is. In the cshtml, the blocks are there, they can be counted, queried and worked on but they don't get displayed unless you reload the page. But if you reload the page multiple times, you get the block multiple times.
I think this has to do with how CSHTML files are pre-processed and/or cached.

The whole idea is not to do this in Javascript. The frontpage page class just inherits from Page, no changes, nothing to add, it is just there so I can get an URL for the front page so I could add the most popular pages or some other links.

Hope there is a quick and easy solution.

**Comment by tidyui** [comment]
Hi there! I see a couple of problems here.

First of let's discuss why the blocks appear *multiple times* when you reload. If you're using memory caching (which is basic in the template setup), the call to `_loader.GetPageAsync(...)` will end up pulling the page from the memory cache (after the first roundtrip to the database). Then you programmatically add blocks to the page which will alter the cached version as it's just returning the cached version from memory. This means that the next time the code runs you get the same object from cache (with the blocks) and add them again, over and over.

Secondly, the call to `_api.Pages.GetAllAsync(...)` with pull *ALL* pages from the database, through model transformation, without going through cache (since only models that are requested specifically with id/slug/etc are put into cache). This is a huge resource waste when you only want three pages and throw the rest away.

## Suggested solution

First of I would put a property `IList<Page> Interests { get; set; }` on your class `FrontPage` without adding any `RegionAttribute`. This means that the property won't be handled or persisted by Piranha. I would then store the pages in this property and render them in the view. 

Secondly I would do a query on the `DbContext` `Piranha.Data.EF.IDB` (https://github.com/PiranhaCMS/piranha.core/blob/master/data/Piranha.Data.EF/IDb.cs) to just get the Id's of the pages I would want to load, then load these pages with `_api.Pages.GetByIdsAsync(...)` (https://github.com/PiranhaCMS/piranha.core/blob/master/core/Piranha/Services/IPageService.cs#L120) and store them in above property. Calling the `GetByIdsAsync` or `GetByIdAsync` ensures the pages will be cached and the next time you retrieve the same pages it won't need to fetch them from the database nor transform the data.

Regards

**Reply by traustitj** [reply]
The code above is mostly just for show, almost pseudo code, that is why there is get all pages and such, just wanted to get that out :)

I will see how far I will get with your suggestions.
